spices/SPICE-0010-overhauled-mapping-listing-typechecks.adoc

```diff
@@ -0,0 +1,432 @@
+  some good fact ✅
+----
+
+Currently, this report is not possible to produce, because reading the top-level `facts` itself will throw the nested member's exception.
+
+== Proposed Solution
+
+To address this, we will defer typechecking of `Mapping` and `Listing` members to when the members are evaluated.
+
+Here is a sample program, and the steps taken by the interpreter:
+
+[source,pkl]
+----
+class Person { name: String }
+
+local people: Listing<Person> = new {
+  new { name = "Sandra" }
+  new { name = "Zoe" }
+}
+
+numPeople = people.length // <1>
+
+sandra = people[0] // <2>
+----
+<1> Neither Sandra nor Zoe are evaluated; yields `2`.
+<2> Sandra is evaluated, and checked against type `Person`. Zoe is not evaluated. Yields `new Person { name = "Sandra" }`.
+
+Here is a more poignant example, where second member is simply a `throw`.
+
+[source,pkl]
+----
+class Person { name: String }
+
+local people: Listing<Person> = new {
+  new { name = "Sandra" }
+  throw("uh oh")
+}
+
+numPeople = people.length
+
+sandra = people[0]
+----
+
+Yet, Pkl still successfully evaluates this program, because the second member is not in the critical path of evaluation.
+
+This principle is consistent with how Pkl behaves generally.
+The property `local foo = throw("uh oh")` similarly does not affect the result of a program, as long as `foo` is never referenced.
+
+== Detailed design
+
+The typecheck of a Mapping and Listing will _type cast_ the value.
+
+Here is a source snippet:
+
+[source,pkl]
+----
+people: Listing<Person> = new {
+  new { name = "Sandra" }
+  new { name = "Zoe" }
+}
+----
+
+And here is the same snippet with the implied instructions:
+
+[source,pkl]
+----
+people: Listing<Person> = new {
+  new { name = "Sandra" } as Person // <1>
+  new { name = "Zoe" } as Person
+} as Listing // <2>
+----
+<1> A type cast is added to each member.
+<2> The overall result is casted to `Listing`.
+
+=== Surrogate objects
+
+Underneath the hood, casting a listing or mapping returns a _new_ listing/mapping.
+
+This new object is a surrogate of the original object.
+It behaves just like the original, except its members receive type casts too.
+This type cast may or may not throw an error.
+
+Given this:
+
+[source,pkl]
+----
+listing = new Listing { "hello" }
+
+listingOfNumbers: Listing<Int> = listing
+----
+
+`"hello"` is not an `Int`, so this program should fail.
+Because typechecking is deferred, this failure should not occur until `listing[0]` is evaluated.
+
+Effectively, `listing` and `listingOfNumbers` are _different_ in-memory objects; one which has type assertions at the end of each member, and one which does not.
+This difference is opaque to users; these two objects have the same hash code, and same equality semantics.
+
+In the amends chain, `listingOfNumbers` has the same parent as `listing`. It delegates its member lookups to `listing`, and adds a typecheck at the end of each lookup.
```

Comment:
@bioball Isn’t this an undesirable breaking change? If I’m not mistaken, introducing a delegation step means that overridden values are no longer reflected in dependent values. Same for the `as` operator.

Reply:
Not totally sure what you mean. Can you provide a snippet that you think is breaking?

BTW: this PR is already merged into main, so feel free to play around with it and see if there is a regression here in Pkl's late-binding semantics.

Reply:
There's an optimization that we make here: if the target type results in the same typecheck, it doesn't do any delegation. So, in this case, there's still just one mapping. See https://github.com/apple/pkl/blob/a7cc09892533b12a80c434937544406b18ae24b2/pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java#L1621-L1623 for implementation details!

Reply:
This is the delegation that's happening:

```
m1: Mapping = new {
  ["name"] = "Bob"
  ["name2"] = this["name"]
}

m2: Mapping<String, String> = (m1) { ["name"] = "Marley"  }

// if m2["name2"] simply delegated to m1["name2"], I'd expect "Bob" here
name2 = m2["name2"] // "Marley"
```

Reply:
I'm still trying to wrap my head around how the delegation works:

```
m1: Mapping = new {
  ["name"] = "Bob"
  ["name2"] = this["name"]
}

// apparently, `m2 = m1 as Mapping<String, String>` isn't equivalent
// to the following. Otherwise, late binding would break (see m3).
m2 = new Mapping {
  ["name"] = m1["name"] as String 
  ["name2"] = m1["name2"] as String
}

m3 = (m2) {
  ["name"] = "Marley"  
}
```

Reply:
Don't know if this is any more helpful, but... the lookup on `name` isn't `m1["name"] as String`; it's `((m1) { ["name"] = "Marley" })["name"] as String`.

Reply:
After checking the implementation, I think I understand what's going on:
The new object resulting from `m1 as Mapping<String, String>` doesn't really define new members that delegate to `m1`'s members.
Instead, the new object knows its value type, and `as String` happens as part of member access.

Reply:
Yup, that's right!